`change` event for input[type=file] does not fire when different file with same name is selected

rr-codes opened the pull request:
#### 5374937e8566ca95b1adefe78fc7b2d90f6497ee
<pre>
`change` event for input[type=file] does not fire when different file with same name is selected
<a href="https://bugs.webkit.org/show_bug.cgi?id=187084">https://bugs.webkit.org/show_bug.cgi?id=187084</a>
rdar://98460303

Reviewed by Ryosuke Niwa.

When a different file with the same name as the currently selected file is selected,
the `change` event does not currently fire, because a file&apos;s path is it&apos;s
identifier, so the two files are considered the same by the system.

This PR works around this by saving each file&apos;s unique id and then comparing the files
both based on their name and now their id too.

* Source/WebCore/fileapi/File.cpp:
(WebCore::File::create):
(WebCore::File::File):
* Source/WebCore/fileapi/File.h:
* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::setFiles):
* Source/WebCore/platform/FileChooser.cpp:
(WebCore::FileChooser::chooseFiles):

Canonical link: <a href="https://commits.webkit.org/255209@main">https://commits.webkit.org/255209@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c025e3cf1df87f49900eb2cd8823e23bfaacc680

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91725 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/958 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22315 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101424 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161494 "Hash c025e3cf for PR 4943 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95730 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/956 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84042 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97730 "Hash c025e3cf for PR 4943 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97383 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/567 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27522 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82491 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/70561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/83151 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35843 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/78196 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33597 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17244 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26991 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3608 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37443 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/80807 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39345 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36361 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17695 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/platform/FileChooser.cpp:
```diff
@@ -58,10 +58,6 @@ void FileChooser::chooseFile(const String& filename)
 
 void FileChooser::chooseFiles(const Vector<String>& filenames, const Vector<String>& replacementNames)
 {
-    // FIXME: This is inelegant. We should not be looking at settings here.
```
the removal of this is necessary since the comparison of just file paths is no longer sufficient

hortont424 replied:
Do you ... know why it was here?

hortont424 commented on Source/WebCore/fileapi/File.cpp:
```diff
@@ -45,7 +45,7 @@ Ref<File> File::createWithRelativePath(ScriptExecutionContext* context, const St
 }
 
-Ref<File> File::create(ScriptExecutionContext* context, const String& path, const String& replacementPath, const String& nameOverride)
+Ref<File> File::create(ScriptExecutionContext* context, const String& path, const String& replacementPath, const String& nameOverride, const std::optional<int64_t>& lastModified)
```
Why is this a const reference instead of just a optional<int64_t>?

rr-codes replied:
because in the `deserialize` it's a const reference so I figured I would do the same; I can change it to not be a const ref if you want?

hortont424 replied:
Which was from Chris in https://trac.webkit.org/changeset/232043/webkit. No, you can leave it.

rniwa commented on Source/WebCore/html/FileInputType.cpp:
```diff
@@ -371,7 +371,7 @@ void FileInputType::setFiles(RefPtr<FileList>&& files, RequestIcon shouldRequest
     else {
         for (unsigned i = 0; i < length; ++i) {
-            if (files->file(i).path() != m_fileList->file(i).path()) {
+            if (files->file(i).path() != m_fileList->file(i).path() || files->file(i).lastModifiedOverride() != m_fileList->file(i).lastModifiedOverride()) {
```
Doesn't this mean that we'd consider two files of the same name & same last modified date as identical?
Why not use the unique identifier of the filesystem instead?
This is stat.st_ino in POSIX, and [FileId in FILE_ID_BOTH_DIR_INFO](https://learn.microsoft.com/en-us/windows/win32/api/winbase/ns-winbase-file_id_both_dir_info) for Windows. [G_FILE_ATTRIBUTE_UNIX_INODE](https://www.freedesktop.org/software/gstreamer-sdk/data/docs/2012.5/gio/GFileInfo.html#G-FILE-ATTRIBUTE-STANDARD-NAME:CAPS) for glib subsystem.

rr-codes replied:
I didn't know those existed! How would I use the Windows and glib functions? (I haven't written many platform specific functions yet)

rniwa replied:
You use GetFileInformationByHandleEx to get the information. See `fileSize(PlatformFileHandle fileHandle)` in FileSystemWin.cpp for an example of calling a similar function: GetFileInformationByHandle

rniwa reviewed (approved):
r=me

rr-codes committed:
`change` event for input[type=file] does not fire when different file with same name is selected
https://bugs.webkit.org/show_bug.cgi?id=187084
rdar://98460303

Reviewed by Ryosuke Niwa.

When a different file with the same name as the currently selected file is selected,
the `change` event does not currently fire, because a file's path is it's
identifier, so the two files are considered the same by the system.

This PR works around this by saving each file's unique id and then comparing the files
both based on their name and now their id too.

* Source/WebCore/fileapi/File.cpp:
(WebCore::File::create):
(WebCore::File::File):
* Source/WebCore/fileapi/File.h:
* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::setFiles):
* Source/WebCore/platform/FileChooser.cpp:
(WebCore::FileChooser::chooseFiles):

Canonical link: https://commits.webkit.org/255209@main
